### ISSUE TITLE
Replace legacy lifecycle methods

### ISSUE THREAD

**Issue by boygirl** [FormidableLabs/victory#969]
replace: 

`componentWillMount`
`componentWillReceiveProps`
`componentWillUpdate`

See https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html for more detail

**Comment by boygirl** [comment]
This issue is complicated by `VictoryAnimation` which relies heavily on `componentWillReceiveProps`. This will require an extensive re-write of Victory's animation strategy.

**Comment by cullenjett** [comment]
I've been playing around with this for a couple of days, and I'm feeling pretty humbled and defeated by it 🏳️. As mentioned there's a good chunk of `componentWillReceiveProps` logic surrounding animations in both the `<VictoryAnimation />` component and even some of the components higher up in the tree (`<VictoryChart />`, `<VictoryStack />`, `<VictoryGroup />`).

@boygirl I agree that a refactor might be necessary in order to fully migrate the lifecycle methods, but I'm not familiar/comfortable enough with the codebase to knock it out. I'd be happy to submit a PR changing the deprecated methods to their `UNSAFE_*` variants and adding `react-lifecycles-compat` until someone smarter can solve the root issue if that adds value?

**Comment by peacechen** [comment]
@cullenjett 
I've found that replacing `componentWillReceiveProps` with `componentDidUpdate` to be straight-forward.  Would you point me to a part of the code that you're trying to refactor?  I could give it a shot for you to review.

**Comment by cullenjett** [comment]
@peacechen Excellent, thanks!

[Here's](https://github.com/FormidableLabs/victory/blob/master/packages/victory-core/src/victory-animation/victory-animation.js#L77) the core `cWRP` call that handles the animation loop. From first glance it seems like a quick rename to `cDU` and changing references to `nextProps` to `this.props` would do the trick, but that prevents the data (bar, line, etc.) from rendering in the chart at all. I didn't dig in on this one much, I spent most of my time being defeated by the next one.

[This one](https://github.com/FormidableLabs/victory/blob/master/packages/victory-chart/src/victory-chart.js#L77) seems to be a candidate for `static gDSFP` since it sets state based on props. I was able to modify the `Wrapper.setAnimationState()` helper function to return a state object instead of setting it directly and using that as the return from `static gDSFP`, but that required some fairly deep code changes that I wasn't confident in and more importantly resulted in subtle animation jank/weird behavior (which is actually pretty similar to what happens if you delete the existing `cWRP` method entirely and don't introduce new lifecycles).

**Comment by boygirl** [comment]
@cullenjett I think your suggested `react-lifecycles-compat` is a really good temporary fix, and definitely helpful. Please do PR this fix.

**Comment by cullenjett** [comment]
@boygirl As I was writing my PR notes I started second-guessing my rationale for adding `react-lifecycles-compat`, so I did some more research -- it seems the `UNSAFE_*` methods aren't actually polyfilled (see [this issue](https://github.com/reactjs/react-lifecycles-compat/issues/2)). 

None of my changes implemented the new lifecycle methods, only changed to `UNSAFE_*` (and removed one or two `cWM` methods), so since I wasn't able to refactor out `componentWillReceiveProps` I'm wondering if there's any value in changing to `UNSAFE_cWRP`? When running in strict mode, React still logs warnings for the `UNSAFE_*` methods, so it seems like there might not be any benefit until we can fully rip out `cWRP`. Does that seem right?
